src/autoschedulers/anderson2021/AutoSchedule.cpp

```diff
@@ -0,0 +1,821 @@
+#include "NetworkSize.h"
+#include "ParamParser.h"
+#include "PerfectHashMap.h"
+#include "State.h"
+
+#ifdef _WIN32
+#include <io.h>
+#define _isatty isatty;
+#else
+#include <unistd.h>
+#endif
+
+namespace Halide {
+namespace Internal {
+namespace Autoscheduler {
+
+using std::string;
+using std::vector;
+
+struct ProgressBar {
+    void set(double progress) {
+        if (!draw_progress_bar) {
+            return;
+        }
+        auto &os = aslog(ProgressBarLogLevel).get_ostream();
+        counter++;
+        const int bits = 11;
+        if (counter & ((1 << bits) - 1)) {
+            return;
+        }
+        const int pos = (int)(progress * 78);
+        os << "[";
+        for (int j = 0; j < 78; j++) {
+            if (j < pos) {
+                os << ".";
+            } else if (j - 1 < pos) {
+                os << "/-\\|"[(counter >> bits) % 4];
+            } else {
+                os << " ";
+            }
+        }
+        os << "]";
+        for (int j = 0; j < 80; j++) {
+            os << "\b";
+        }
+    }
+
+    void clear() {
+        if (counter) {
+            auto &os = aslog(ProgressBarLogLevel).get_ostream();
+            for (int j = 0; j < 80; j++) {
+                os << " ";
+            }
+            for (int j = 0; j < 80; j++) {
+                os << "\b";
+            }
+        }
+    }
+
+private:
+    uint32_t counter = 0;
+    static constexpr int ProgressBarLogLevel = 1;
+    const bool draw_progress_bar = isatty(2) && aslog::aslog_level() >= ProgressBarLogLevel;
+};
+
+// TODO: this is scary as heck, can we be sure all these references don't go stale?
+struct AutoSchedule {
+    const FunctionDAG &dag;
+    Anderson2021Params params;
+    const Target &target;
+    const std::vector<Function> &outputs;
+    std::mt19937 &rng;
+    CostModel *cost_model;
+    Statistics &stats;
+    SearchSpace &search_space;
+    const LoopNestParser *partial_schedule;
+
+    AutoSchedule(const FunctionDAG &dag,
+                 const Anderson2021Params &params,
+                 const Target &target,
+                 const std::vector<Function> &outputs,
+                 std::mt19937 &rng,
+                 CostModel *cost_model,
+                 Statistics &stats,
+                 SearchSpace &search_space,
+                 const LoopNestParser *partial_schedule);
+
+    bool use_partial_schedule() const {
+        return partial_schedule;
+    }
+
+    IntrusivePtr<State> optimal_schedule_pass(int beam_size,
+                                              int pass_idx,
+                                              int num_passes,
+                                              ProgressBar &tick,
+                                              std::unordered_set<uint64_t> &permitted_hashes);
+
+    // Performance coarse-to-fine beam search and return the best state found.
+    IntrusivePtr<State> optimal_schedule(int beam_size);
+};
+
+#ifdef HALIDE_ALLOW_LEGACY_AUTOSCHEDULER_API
```

Comment:
FYI: HALIDE_ALLOW_LEGACY_AUTOSCHEDULER_API is now removed in Halide `main`, so this should just be removed here and elsewhere. (We can do this as a followup PR if you like.)